ENIGMAsystem/SHELL/Bridges/General/Win32-Direct3D11.cpp

```diff
@@ -122,105 +122,156 @@ void initialize_render_targets() {
+  swapChainDesc.SampleDesc.Count = 1;
+  swapChainDesc.SampleDesc.Quality = D3D11_STANDARD_MULTISAMPLE_PATTERN;
+  swapChainDesc.OutputWindow = enigma::hWnd;
+  swapChainDesc.Windowed = true; // initially windowed and not fullscreen
+  swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_DISCARD;
+  swapChainDesc.Flags = 0;
+
+  D3D_FEATURE_LEVEL featureLevel = D3D_FEATURE_LEVEL_11_0;
+
+  UINT flags = 0;
+  #ifdef DEBUG_MODE
+  flags |= D3D11_CREATE_DEVICE_DEBUG;
+  #endif
+  result = D3D11CreateDeviceAndSwapChain(NULL, D3D_DRIVER_TYPE_HARDWARE, NULL, flags, &featureLevel, 1,
+                                          D3D11_SDK_VERSION, &swapChainDesc, &m_swapChain, &m_device, NULL, &m_deviceContext);
+  if (FAILED(result)) {
+    show_error("Failed to create Direct3D11 device and swap chain.", true);
+  }
+
+  initialize_render_targets();
+
+  // Setup the raster description which will determine how and what polygons will be drawn.
+  D3D11_RASTERIZER_DESC rasterDesc = { };
+  rasterDesc.AntialiasedLineEnable = false;
+  rasterDesc.CullMode = D3D11_CULL_NONE;
+  rasterDesc.DepthBias = 0;
+  rasterDesc.DepthBiasClamp = 0.0f;
+  rasterDesc.DepthClipEnable = false;
+  rasterDesc.FillMode = D3D11_FILL_SOLID;
+  rasterDesc.FrontCounterClockwise = false;
+  rasterDesc.MultisampleEnable = false;
+  rasterDesc.ScissorEnable = false;
+  rasterDesc.SlopeScaledDepthBias = 0.0f;
+
+  result = m_device->CreateRasterizerState(&rasterDesc, &m_rasterState);
+  if (FAILED(result)) {
+    show_error("Failed to create Direct3D11 rasterizer state.", true);
   }
 
-  void DisableDrawing(void* handle) {}
+  m_deviceContext->RSSetState(m_rasterState);
 
-  void ScreenRefresh() {
-    m_swapChain->Present(swap_interval, 0);
+  enigma_user::display_aa = 0;
+  UINT quality_levels = 0;
+  for (int i = 16; i > 1; i--) {
+    if (SUCCEEDED(m_device->CheckMultisampleQualityLevels(swapChainDesc.BufferDesc.Format, i, &quality_levels))) {
+      if (quality_levels > 0)
+        enigma_user::display_aa += i;
+    }
   }
 }
 
-namespace enigma_user
-{
-  int display_aa = 0;
+void DisableDrawing(void* handle) {}
 
-  void display_reset(int samples, bool vsync) {
-    swap_interval = vsync ? 1 : 0;
-  }
+void ScreenRefresh() {
+  m_swapChain->Present(swap_interval, 0);
+}
 
-  void set_synchronization(bool enable)
-  {
-    swap_interval = enable ? 1 : 0;
-  }
+} // namespace enigma
+
+namespace enigma_user {
+
+int display_aa = 0;
+
+void display_reset(int samples, bool vsync) {
+  swap_interval = vsync ? 1 : 0;
+
+  // when disabling msaa, we still need at least 1 samples
+  if (samples < 1) samples = 1;
+
+  // get the old swap chain description and turn on msaa
+  DXGI_SWAP_CHAIN_DESC swapChainDesc;
+  m_swapChain->GetDesc(&swapChainDesc);
+
+  // short-circuit if we already have a swap chain with this many samples
+  if (swapChainDesc.SampleDesc.Count == samples) return;
+
+  // release all the old references first
+  m_deviceContext->OMSetRenderTargets(0, 0, 0);
```

Comment:
For instance, here...

Reply:
Yeah, it's just these two places right now that we need to do something like this. We are releasing the same things in both cases, but in the second case we are additionally releasing the swap chain too. Are you trying to say this is enough to warrant a helper function now?

Reply:
Absolutely. Even one place for something like a list of calls to some five `deinit` or `free` or `release` methods is cause for a helper.

Reply:
Ok, I'll take care of it.

Reply:
Here we go: 3aca962bbf8e6d9af5d35d55bcedae95c99dcf8b

I'll start taking up the practice of doing this from here on out. I mean, I will lower my threshold for what I consider worthy of a helper.